### PR TITLE
Refactor: Remove CustomAPI routes for BTC and ETH

### PR DESCRIPTION
**Pull request by catenocrypt**
Issue #1019 .

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -5,16 +5,24 @@ import (
...
 )
 
-func SetupPlatformAPI(root gin.IRouter) {
-	for _, api := range platform.Platforms {
-		RegisterCollectionsAPI(root, api)
-		RegisterTransactionsAPI(root, api)
-		RegisterCustomAPI(root, api)
-		RegisterTokensAPI(root, api)
-		RegisterStakeAPI(root, api)
+func SetupPlatformAPI(router gin.IRouter) {
```
Comment (EnoRage) [inline api/api.go:8]: What is the purpose to split it?

File: pkg/blockatlas/platform.go [hunk 2]
```diff
@@ -23,23 +29,16 @@ type (
...
-		GetTokenListByAddress(address string) (TokenPage, error)
-	}
-
-	// BlockAPI provides block information and lookups
-	BlockAPI interface {
-		Platform
-		CurrentBlockNumber() (int64, error)
-		GetBlockByNumber(num int64) (*Block, error)
+	// TxByAddrAndXPubAPI provides transaction lookup based on address and XPUB (Bitcoin-style)
+	TxByAddrAndXPubAPI interface {
```
Comment (EnoRage) [inline pkg/blockatlas/platform.go:33]: XpubAPI

**Comment by EnoRage** [comment]
Great! Will add couple comments more soon

File: api/api.go (hunk 1 above)
Comment (catenocrypt) [reply api/api.go:8]: A small step in the direction when services are more separated, and handlers are not decided based on interface implementation or not, but explicit lists.  Staking and Naming (as there are only a few of them) have been completely separated.
I plan to change the routes one for each coin, then the for loops here will be gone.

Comment (EnoRage) [reply api/api.go:8]: Cannot see any sense. You are just make more code by making 5 loops instead of 1

File: pkg/blockatlas/errors.go [hunk 3]
```diff
@@ -10,3 +10,6 @@ var ErrInvalidAddr = errors.New("invalid address")
 
 // ErrNotFound signals that the resource has not been found
 var ErrNotFound = errors.New("not found")
+
+// ErrInvalidKey signals that the requested key is invalid
+var ErrInvalidKey = errors.New("invalid key")
```
Comment (prazd) [inline pkg/blockatlas/errors.go:15]:
```
var ( 
   ErrNotFound
   ErrInvalidKey
   ...
)
```

File: platform/ethereum/transaction.go [hunk 4]
```diff
@@ -1,45 +1,18 @@
...
+func (p *Platform) GetTokenTxsByAddress(address string, token string) (blockatlas.TxPage, error) {
+	page, err := p.client.GetTransactions(address, p.CoinIndex)
 	if err != nil {
-		logger.Error(err, "Unhandled error")
-		c.AbortWithStatus(http.StatusInternalServerError)
-		return true
+		return nil, err
 	}
-	return false
+	return page, err
```
Comment (prazd) [inline platform/ethereum/transaction.go:17]: Maybe return `page, nil`?
Or you can delete `if err != nil` check:
```
return p.client.GetTransactions(address, p.CoinIndex)
```

File: platform/platform.go [hunk 5]
```diff
@@ -111,7 +109,22 @@ func getPlatformMap() blockatlas.Platforms {
 	}
 }
 
-func InitCollectionsWhitelist() {
-	CollectionsWhitelist = make(map[uint]bool)
-	CollectionsWhitelist[coin.Ethereum().ID] = true
+func getCollectionsHandlers() map[uint]blockatlas.CollectionsAPI {
```
Comment (prazd) [inline platform/platform.go:112]: Maybe add alias for this map?

File: api/registry.go [hunk 6]
```diff
@@ -6,143 +6,88 @@ import (
...
-	root.GET("/ns/lookup", endpoint.GetAddressByCoinAndDomain)
-	root.GET("v2/ns/lookup", endpoint.GetAddressByCoinAndDomainBatch)
-}
-
-func RegisterBasicAPI(root gin.IRouter) {
-	root.GET("/", endpoint.GetStatus)
-	root.GET("/metrics", ginprom.PromHandler(promhttp.Handler()))
+func RegisterDomainAPI(router gin.IRouter) {
+	router.GET("/ns/lookup", endpoint.GetAddressByCoinAndDomain)
+	router.GET("v2/ns/lookup", endpoint.GetAddressByCoinAndDomainBatch)
```
Comment (prazd) [inline api/registry.go:87]: `/v2/ns/lookup/`

**Comment by catenocrypt** [comment]
I planned to make API routes generic w.r.t. coin, e.g. '/v1/:coin/address/:address' instead of '/v1/' + coin + '/address/:address'.  However, this is not easily possible.  There are many routes that would conflict this way:

- "/v1/:coin/:address"
- "/v1/:coin/address/:address"

and
- "/v2/:coin/transactions/:address"
- "/v2/:coin/staking/validators"
- "/v2/staking/delegations"
- etc.

The only way to make this is through an overly generic API path like "/v1/:arg1/:arg2/:arg3", and dispatch from code.  But this would be not clean, hard to see, hard to document, etc.

 Therefore the correct scheme of per-coin API routes remains
`for (handler := ...)  "/v1/" + handler.coin + "/address/:address/"`

File: api/api.go (hunk 1 above)
Comment (catenocrypt) [reply api/api.go:8]: It's about separation of concerns.  Think like this:

A: Get a bunch of handlers, DECIDE which ones are relevant and REGISTER an path for them
B: Get a list of relevant handlers, REGISTER a path for them.

Separation of concerns sometimes comes with a few extra lines, and I think it worth a few lines.
However, my plan was to change the routes generic w.r.t. coin, in which case the for loops would have disappear completely, but this turned to be not plausible (see prev. comment).  Therefor I may revert this change.

File: platform/ethereum/transaction.go (hunk 4 above)
Comment (catenocrypt) [reply platform/ethereum/transaction.go]: yes, cleaner, thanks

File: api/registry.go (hunk 6 above)
Comment (catenocrypt) [reply api/registry.go]: I added starting / to 3 routes that missed it, but I'm not sure about the trailing slashes, that's missing in many routes.  What's the significance of trailing slash?

File: platform/platform.go (hunk 5 above)
Comment (catenocrypt) [reply platform/platform.go]: done

**Comment by catenocrypt** [comment]
Still keeping as WIP, I need to test it more thoroughly.

**Comment by catenocrypt** [comment]
API unit tests have been verified together with contents; exact same output is produce before PR.

**Comment by catenocrypt** [comment]
This PR is ready for review now, please review.

**Comment by EnoRage** [comment]
Will take a look today

**Comment by EnoRage** [comment]
PR will be merged with ignored codecov/patch check

**Pull request merged**